src/flash-source-buffer.js

```diff
@@ -382,17 +385,26 @@ export default class FlashSourceBuffer extends videojs.EventTarget {
 
       if (startTag.pts >= targetPts) {
         filteredVideoTags.push(startTag);
-      } else if (startTag.keyFrame) {
-        let nextIndex = startIndex + 1;
+      } else if (startTag.metaDataTag || startTag.keyFrame) {
+        let keyFrameIndex = getKeyFrameIndex(startTag, startIndex);
+
+        if (!verifyKeyFrameIndex(videoTags, keyFrameIndex)) {
+          startIndex++;
+          continue;
+        }
+
+        let nextIndex = keyFrameIndex + 1;
         let foundNextKeyFrame = false;
 
         while (nextIndex < videoTags.length) {
```

Comment:
This whole thing really needs some comments. What is the purpose of this inner loop? The outer loop?